src/libopensc/card-starcos.c

```diff
@@ -281,7 +292,64 @@ static unsigned int starcos_determine_pin_encoding(sc_card_t *card)
 	return encoding;
 }
 
+/**
+ * Returns 1 if an extended APDU can be sent to the card
+ * with the given card reader. Otherwise returns 0.
+ */
+static int starcos_probe_reader_for_ext_apdu(sc_card_t * card) {
+	sc_apdu_t apdu;
+	int rv;
+	/* try to read STARCOS3X_PROBE_APDU_LENGTH bytes */
+	u8 data[STARCOS3X_PROBE_APDU_LENGTH];
+
+	/* Get Data: Get Chip Serial Number */
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_EXT, 0xCA, 0x9F, 0x6C);
+	apdu.cla = 0xA0;
+	apdu.resp = data;
+	apdu.resplen = sizeof(data);
+	apdu.le = apdu.resplen;
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rv, "Failed to send Get Data ext. APDU");
+	return (apdu.sw1 == 0x90 && apdu.sw2 == 0x00);	
+}
+
+static int starcos_select_mf(sc_card_t * card) {
+	sc_apdu_t apdu;
+	const u8 mf_buf[2] = {0x3f, 0x00};
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0xA4, 0x00, 0x0C);
+	apdu.le = 0;
+	apdu.lc = 2;
+	apdu.data    = mf_buf;
+	apdu.datalen = 2;
+	apdu.resplen = 0;
+	
+	return sc_transmit_apdu(card, &apdu);
+}
+
+static int starcos_select_aid(sc_card_t *card,
+			      const u8 aid[16], size_t len,
+			      sc_file_t **file_out);
+
+/* returns 1 if the card has the eSign app with AID A0:00:00:02:45:53:69:67:6E
+   otherwise returns 0
+ */
+static int starcos_has_esign_app(sc_card_t * card) {
+	
+	static u8 esign_aid[SC_MAX_PATH_SIZE] = { 0xA0, 0x00, 0x00, 0x02, 0x45, 0x53, 0x69, 0x67, 0x6E };
+	int rv;
 
+	rv = starcos_select_mf(card);
+	if ( rv == SC_SUCCESS ) {
+		sc_file_t * file;
+		rv = starcos_select_aid(card, esign_aid, 9, &file);
```

Comment:
please, try to avoid using magic numbers:
```suggestion
		rv = starcos_select_aid(card, esign_aid, sizeof esign_aid, &file);
```